### PR TITLE
Preview refactoring

### PR DESCRIPTION
**Pull request by diegocepedaw**
Change preview to use temporary tables and the existing populate function.

### CODE REVIEW

File: src/oncall/scheduler/default.py [hunk 1]
```diff
@@ -501,12 +448,20 @@ def populate(self, schedule, start_time, dbinfo):
...
                 continue
-            self.create_events(team_id, schedule['id'], user_id, epoch, role_id, cursor)
-        connection.commit()
\ No newline at end of file
+            self.create_events(team_id, schedule['id'], user_id, epoch, role_id, cursor, table_name)
+
+        # if this function is being called by preview return events
+        if req.method == 'GET':
+            resp.body = self.build_preview_response(req, resp, cursor, table_name)
+            query = "DROP TABLE %s;" % table_name
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: We'd want to interpolate the table name the same way we do other queries, as arguments to `cursor.execute`

File: src/oncall/scheduler/default.py [hunk 2]
```diff
@@ -486,6 +426,13 @@ def populate(self, schedule, start_time, dbinfo):
         handoff = start_epoch + timedelta(seconds=first_event_start)
         handoff = timezone(schedule['timezone']).localize(handoff)
 
+        # create temporary table if function is being used to preview
+        if req.method == 'GET':
+            start__lt = req.get_param('start__lt', required=True)
+            end__ge = req.get_param('end__ge', required=True)
+            query = 'CREATE TEMPORARY TABLE IF NOT EXISTS `%s` AS (SELECT * FROM `event` WHERE `start` < %s AND `end` > %s)' % (table_name, start__lt, end__ge)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: we want to interpolate these values using arguments cursor.execute to prevent sql injection problems

File: src/oncall/scheduler/default.py (hunk 1 above)
Comment (diegocepedaw) [reply src/oncall/scheduler/default.py]: There is no reason the table name needs to be interpolated here instead of just being hard coded so I changed it to a hard coded query.

File: src/oncall/scheduler/default.py (hunk 2 above)
Comment (diegocepedaw) [reply src/oncall/scheduler/default.py]: Hard coded the table name and changed the interpolation to use the cursor.execute function.

File: src/oncall/scheduler/default.py [hunk 3]
```diff
@@ -501,12 +447,20 @@ def populate(self, schedule, start_time, dbinfo):
...
                 continue
-            self.create_events(team_id, schedule['id'], user_id, epoch, role_id, cursor)
-        connection.commit()
\ No newline at end of file
+            self.create_events(team_id, schedule['id'], user_id, epoch, role_id, cursor, table_name)
+
+        # if this function is being called by preview return events
+        if req.method == 'GET':
+            resp.body = self.build_preview_response(req, resp, cursor, table_name)
+            query = "DROP TABLE `temp_event`"
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: This seems to ignore your custom `table_name` variable. The change i had in mind was:

    query = "DROP TABLE %s"
    cursor.execute(query, [table_name])

Comment (diegocepedaw) [reply src/oncall/scheduler/default.py]: It does indeed ignore it. Revisiting it, it didn't seem necessary to use the variable since only preview uses GET and preview always uses the table name 'temp_event'. In fact, 'temp_event' is already hardcoded within the columns and constraints. This way seemed bit more readable and unlikely to drop the original 'event' database if for some reason a GET endpoint for populate was added.

Comment (jrgp) [reply src/oncall/scheduler/default.py]: ah gotcha. should probably remove all references to the `table_name` variable throughout the code/PR for consistency if it won't be used

Comment (diegocepedaw) [reply src/oncall/scheduler/default.py]: So table_name is used when creating events and deleting events (since both preview and populate use the same functions), it's just the creating and dropping of the temporary table that  don't need a variable since they are unique to preview.

File: src/oncall/scheduler/default.py [hunk 4]
```diff
@@ -123,50 +123,52 @@ def get_busy_user_by_event_range(self, user_ids, team_id, events, cursor):
...
-            SELECT `user_id`, MAX(`end`) AS `last_end` FROM `event`
-            WHERE `team_id` = %s AND `user_id` IN %s AND `end` <= %s
+        user_id_list = ', '.join(map(str, user_ids))
+        query = '''
+            SELECT `user_id`, MAX(`end`) AS `last_end` FROM `%s`
+            WHERE `team_id` = %s AND `user_id` IN (%s) AND `end` <= %s
             AND `role_id` = %s
             GROUP BY `user_id`
-            ''', (team_id, user_ids, start_time, role_id))
+            ''' % (table_name, team_id, user_id_list, start_time, role_id)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/default.py [hunk 5]
```diff
@@ -123,50 +123,52 @@ def get_busy_user_by_event_range(self, user_ids, team_id, events, cursor):
...
             JOIN `user` ON `user`.`id` = `roster_user`.`user_id` AND `roster_user`.`roster_id` = %s
-            LEFT JOIN `event` ON `event`.`user_id` = `user`.`id` AND `event`.`team_id` = %s AND `event`.`end` <= %s
-                AND `event`.`role_id` = %s
-            WHERE `roster_user`.`in_rotation` = 1 AND `event`.`id` IS NULL
-        '''
-        cursor.execute(query, (roster_id, team_id, start_time, role_id))
+            LEFT JOIN `%s` ON `%s`.`user_id` = `user`.`id` AND `%s`.`team_id` = %s AND `%s`.`end` <= %s
+                AND `%s`.`role_id` = %s
+            WHERE `roster_user`.`in_rotation` = 1 AND `%s`.`id` IS NULL
+        ''' % (roster_id, table_name, table_name, table_name, team_id, table_name, start_time, table_name, role_id, table_name)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/default.py [hunk 6]
```diff
@@ -185,24 +188,24 @@ def create_events(self, team_id, schedule_id, user_id, events, role_id, cursor,
...
             logger.debug('inserting event: %s', event_args)
             query = '''
-                INSERT INTO `event` (
+                INSERT INTO `%s` (
                     `team_id`, `schedule_id`, `start`, `end`, `user_id`, `role_id`
                 ) VALUES (
                     %s, %s, %s, %s, %s, %s
-                )'''
-            cursor.execute(query, event_args)
+                )''' % (table_name, team_id, schedule_id, event['start'], event['end'], user_id, role_id)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/default.py [hunk 7]
```diff
@@ -185,24 +188,24 @@ def create_events(self, team_id, schedule_id, user_id, events, role_id, cursor,
...
                 query = '''
-                    INSERT INTO `event` (
+                    INSERT INTO `%s` (
                         `team_id`, `schedule_id`, `start`, `end`, `user_id`, `role_id`, `link_id`
                     ) VALUES (
-                        %s, %s, %s, %s, %s, %s, %s
-                    )'''
-                cursor.execute(query, event_args)
+                        %s, %s, %s, %s, %s, %s, '%s'
+                    )''' % (table_name, team_id, schedule_id, event['start'], event['end'], user_id, role_id, link_id)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/default.py [hunk 8]
```diff
@@ -501,12 +447,20 @@ def populate(self, schedule, start_time, dbinfo):
         future_events = filter(lambda x: x != [], future_events)
         if future_events:
             first_event_start = min(future_events[0], key=lambda x: x['start'])['start']
-            cursor.execute('DELETE FROM event WHERE schedule_id = %s AND start >= %s', (schedule['id'], first_event_start))
+            query = 'DELETE FROM %s WHERE schedule_id = %s AND start >= %s' % (table_name, schedule['id'], first_event_start)
```
Comment (jrgp) [inline src/oncall/scheduler/default.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/round-robin.py [hunk 9]
```diff
@@ -7,22 +7,24 @@
...
-                        (SELECT `user_id`, MAX(`start`) AS `last_start` FROM `event`
-                         WHERE `team_id` = %s AND `user_id` IN %s AND `start` <= %s
+                        (SELECT `user_id`, MAX(`start`) AS `last_start` FROM `%s`
+                         WHERE `team_id` = %s AND `user_id` IN (%s) AND `start` <= %s
                          AND `role_id` = %s
                          GROUP BY `user_id`
                          ORDER BY `last_start` DESC) t
                         LIMIT 1
-                        ''', (schedule['team_id'], roster, start, schedule['role_id']))
+                        ''' % (table_name, schedule['team_id'], roster_list, start, schedule['role_id'])
```
Comment (jrgp) [inline src/oncall/scheduler/round-robin.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/round-robin.py [hunk 10]
```diff
@@ -42,25 +44,25 @@ def find_next_user_id(self, schedule, future_events, cursor):
...
             logger.debug('inserting event: %s', event_args)
             query = '''
-                INSERT INTO `event` (
+                INSERT INTO `%s` (
                     `team_id`, `schedule_id`, `start`, `end`, `user_id`, `role_id`
                 ) VALUES (
                     %s, %s, %s, %s, %s, %s
-                )'''
-            cursor.execute(query, event_args)
+                )''' % (table_name, team_id, schedule_id, event['start'], event['end'], user_id, role_id)
```
Comment (jrgp) [inline src/oncall/scheduler/round-robin.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/round-robin.py [hunk 11]
```diff
@@ -71,18 +73,12 @@ def create_events(self, team_id, schedule_id, user_id, events, role_id, cursor,
                         `team_id`, `schedule_id`, `start`, `end`, `user_id`, `role_id`, `link_id`
                     ) VALUES (
                         %s, %s, %s, %s, %s, %s, %s
-                    )'''
-                cursor.execute(query, event_args)
+                    )''' % (table_name, team_id, schedule_id, event['start'], event['end'], user_id, role_id, link_id)
```
Comment (jrgp) [inline src/oncall/scheduler/round-robin.py]: Can these be interpolated using cursor.execute?

File: src/oncall/scheduler/default.py (hunk 3 above)
Comment (jrgp) [reply src/oncall/scheduler/default.py]: ah that makes sense

File: src/oncall/scheduler/default.py (hunk 4 above)
Comment (dwang159) [reply src/oncall/scheduler/default.py]: `table_name` can't be, but the rest should be to avoid SQL injection

File: src/oncall/scheduler/default.py [hunk 12]
```diff
@@ -486,6 +426,12 @@ def populate(self, schedule, start_time, dbinfo):
         handoff = start_epoch + timedelta(seconds=first_event_start)
         handoff = timezone(schedule['timezone']).localize(handoff)
 
+        # create temporary table if function is being used to preview
```
Comment (dwang159) [inline src/oncall/scheduler/default.py]: We should limit the scope of the populate function. `populate()` shouldn't be responsible for determining whether the action is a preview or a populate request, and it shouldn't include falcon's `req` or `resp` as arguments. Instead, populate should do exactly that, populate the schedule. 

We can go back to creating a `preview` function, which can handle this logic. Both `preview()` and `populate()` can rely on another helper `schedule_events()` (or something similar), which can generate the populate event data. Then, `preview` can pass this info back to the caller in `api/v0/preview.py` and `populate` can update the DB. The falcon responder in `api/v0/preview.py` can then be responsible for formatting and updating the HTTP response with the schedule preview data.

File: src/oncall/api/v0/populate.py [hunk 13]
```diff
@@ -45,6 +46,6 @@ def on_post(req, resp, schedule_id):
     scheduler = load_scheduler(scheduler_name)
     schedule = get_schedules({'id': schedule_id})[0]
     check_team_auth(schedule['team'], req)
-    scheduler.populate(schedule, start_time, (connection, cursor))
+    scheduler.populate(schedule, start_time, (connection, cursor), req, resp, table_name)
```
Comment (dwang159) [inline src/oncall/api/v0/populate.py]: The temporary table details should be left out of the call to populate/preview, to allow other schedulers flexibility in defining their own populate/preview logic.

File: src/oncall/scheduler/default.py [hunk 14]
```diff
@@ -14,40 +14,40 @@
 SECONDS_IN_A_WEEK = SECONDS_IN_A_DAY * 7
 
 columns = {
-    'id': '`event`.`id` as `id`',
```
Comment (dwang159) [inline src/oncall/scheduler/default.py]: Why is this needed? Preview shouldn't need these constraints. Also, move this to `preview.py`, so all falcon logic is handled in the API responders

File: src/oncall/scheduler/default.py [hunk 15]
```diff
@@ -123,50 +123,52 @@ def get_busy_user_by_event_range(self, user_ids, team_id, events, cursor):
...
+            ''' % (table_name, team_id, user_id_list, start_time, role_id)
+        cursor.execute(query)
         if cursor.rowcount != 0:
             # Grab user id with lowest last scheduled time
             return min(cursor.fetchall(), key=operator.itemgetter('last_end'))['user_id']
         else:
             return None
 
-    def find_new_user_in_roster(self, roster_id, team_id, start_time, role_id, cursor):
+    def find_new_user_in_roster(self, roster_id, team_id, start_time, role_id, cursor, table_name):
```
Comment (dwang159) [inline src/oncall/scheduler/default.py]: Recommend setting this as a default argument: `table_name='event'`
